### PR TITLE
Introduce the BYOD featured notebook.

### PR DESCRIPTION
**Pull request by xbrianh**
This notebook includes application logic to join Terra data tables over a column.

### CODE REVIEW

File: notebooks/byod.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    ## Preview the data in your workspace bucket
+    """
+
+with callysto.Cell("python"):
+    #!gsutil ls {bucket}
+    pass
+
+with callysto.Cell("markdown"):
+    """
+    # Generate a data table that links to the data in your workspace bucket
```
Comment (bethsheets) [inline notebooks/byod.py:186]: Were you planning to add example code here for using your functions above to create a table?

File: notebooks/byod.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+with callysto.Cell("markdown"):
+    """
+    If you are planning to upload many files to your Terra workspace, we recommend you organize your data into a Terra
+    data table. This is especially helpful when you plan to run workflows with this data because you can avoid pasting
+    in the "gs://" links to every file and instead use Terra's helpful UI features.
+
+
+    In this example, we introduce tools to help you:
+
+    1. Programmatically upload data from your local machine to your Terra workspace using `gsutil cp`.
```
Comment (aofarrel) [inline notebooks/byod.py:24]: We definitely duplicated work in the creation of the tables, but this part and mine differ in how the files get into the bucket. If we combine efforts I think this part will go unchanged; it's much better than my sort of handwaved "just get them in the bucket somehow" approach.

File: notebooks/byod.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    def join_data_tables(new_table: str, tables_to_join: list, join_column: str):
+        keyed_rows = get_keyed_rows(tables_to_join[0], join_column)
+        for table_name in tables_to_join[1:]:
+            keyed_rows = join_keyed_rows(keyed_rows, get_keyed_rows(table_name, join_column))
+        upload_rows(new_table, [{join_column: k, **row} for k, row in keyed_rows.items()])
+
+    def create_cram_crai_table(table: str, listing: Iterable[str]):
+        crams = dict()
+        crais = dict()
+        for key in listing:
```
Comment (aofarrel) [inline notebooks/byod.py:126]: I don't think this approach will work if the crai files are formatted as NWD1.cram.crai, which is pretty common. Even though I think the sample names would be correct, I think 
`elif "crai" == ext`
would never fire. I think I used .endswith() but I'll have to double check.

File: notebooks/byod.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    """
+    ### Add a prefix to your bucket path to organize your data
+    In this example, we add the prefix 'test-crai-cram'. In the terminal of your computer, you will call something like:
+
+    `gsutil cp /Users/Documents/Example.cram gs://your_bucket_info/test-crai-cram/`
+    """
+
+with callysto.Cell("markdown"):
+    """
+    ## Preview the data in your workspace bucket
```
Comment (aofarrel) [inline notebooks/byod.py:177]: Warn the user that a nightmarish amount of text will ensue if they do this on a bucket with hundreds of files.

File: notebooks/byod.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+upload_columns("test_metadata_table_a", test_metadata_table_a_columns)
+
+delete_table("test_metadata_table_b")
+test_metadata_table_b_columns = dict(sample=["sample_id_1", "sample_id_2", "sample_id_3", "sample_id_5"],
+                                     alpha=["1", "2", "3", "4"],
+                                     beta=["5", "6", "7", "8"])
+upload_columns("test_metadata_table_b", test_metadata_table_b_columns)
+
+delete_table("test_joined_table")
+join_data_tables("test_joined_table", ["test_cram_crai_table", "test_metadata_table_a", "test_metadata_table_b"], "sample")
```
Comment (aofarrel) [inline notebooks/byod.py:214]: If I understand what you're doing here we've definitely duplicated efforts. Let's see if we can merge our notebooks.

**Review by aofarrel — CHANGES_REQUESTED** [review]
My old TSVafy notebooks heavily overlap with this use case. For the sake of organization and not further doubling efforts, I think we should merge them, even though the approaches they take are quite different.
I think I must've deleted my old repo with them but here's a new one: https://github.com/aofarrel/ttt

File: notebooks/byod.py (hunk 1 above)
Comment (xbrianh) [reply notebooks/byod.py:186]: Yes, will add.

Comment (xbrianh) [reply notebooks/byod.py:187]: example added

File: notebooks/byod.py (hunk 3 above)
Comment (xbrianh) [reply notebooks/byod.py:127]: Thanks, I've added code to cover `foo.cram.crai` filename formats.

File: notebooks/byod.py (hunk 5 above)
Comment (xbrianh) [reply notebooks/byod.py:291]: This is a general method that will work for any Terra data tables with columns of shared values, whether or not they were imported from Gen3.

File: notebooks/byod.py (hunk 4 above)
Comment (xbrianh) [reply notebooks/byod.py:186]: If they don't know, they'll find out. I feel that general knowledge about working with buckets does not necessarily belong in this guide.

**Comment by xbrianh** [comment]
@bethsheets @aofarrel @DailyDreaming I've merged this into master for now. We can keep collaborating on it through additional pull requests.

When we feel it is ready to be published, we can change the [publish target workspace](https://github.com/DataBiosphere/bdcat_notebooks/blob/11e95dd925442f8a5e031a1835228db5894e99b5/notebooks/byod.py#L1) from "terra-notebook-utils-tests" to "BioData Catalyst Collection".

**Pull request merged**